strawberryfields/decompositions.py

```diff
@@ -1070,6 +1070,363 @@ def bloch_messiah(S, tol=1e-10, rounding=9):
+    This sequence of :math:`\mathrm{SU}(2)` transformations can then be mapped to an operation
+    on optical modes including two phase plates and one beam splitter.
+
+    This implementation is based on the authors' code at `github:glassnotes/Caspar
+    <https://github.com/glassnotes/Caspar>`_.
+
+    Args:
+        U (array): unitary matrix
+        rtol (float): relative tolerance used when checking if the matrix is unitary
+        atol (float): absolute tolerance used when checking if the matrix is unitary
+    Returns:
+        tuple[list[tuple,list], float]: Returns a list of operations with elements in
+        the form ``(i,i+1), [a, b, g]`` where the ``(i,i+1)`` indicates the modes of an
+        :math:`\mathrm{SU}(2)` transformation and ``[a, b, g]`` are the transformation parameters.
+
+    .. details::
+
+    Note that any unitary can be written in terms of an special unitary as
+
+    .. math:: U = e^{i \phi/n} S
+
+    where :math:`S \in \mathrm{SU}(n)` and :math:`e^{i\phi} = \mathrm{det}\,U`.
+
+    Here any :math:`S \in \mathrm{SU}(n)` is parametrized in terms of the Euler angles and written as
+
+    .. math::
+        S(\alpha, \beta, \gamma) =
+        \begin{pmatrix}
+            e^{i\alpha/2} & 0              \\
+            0             & e^{-i\alpha/2}
+        \end{pmatrix}
+        \begin{pmatrix}
+            \cos{\beta/2} & -\sin{\beta/2}  \\
+            \sin{\beta/2}  & \cos{\beta/2}
+        \end{pmatrix}
+        \begin{pmatrix}
+            e^{i\gamma/2} & 0              \\
+            0             & e^{-i\gamma/2}
+        \end{pmatrix}.
+
+    This factorization then determines the constructions of the :math:`\mathrm{SU}(2)` device
+    acting on the respective optical modes
+
+    .. math::
+         S(\alpha, \beta, \gamma) =
+            \left[ R(\alpha/2) \otimes R(-\alpha/2) \right] \,
+            BS(\beta/2) \,
+            \left[ R(\gamma/2) \otimes R(-\gamma/2) \right].
+    """
+
+    n = U.shape[0]
+    parameters = []
+    global_phase = None
+    det = np.linalg.det(U)
+
+    if n < 3:
+        raise ValueError("Input matrix for decomposition must be at least 3x3.")
+    if not np.allclose(U @ U.conj().T, np.identity(n), rtol=rtol, atol=atol):
+        raise ValueError("The input matrix is not unitary.")
+
+    # if Unitary, factorize into phase times Special Unitary
+    SU = U.copy()
+    if not np.isclose(det, 1, rtol=rtol, atol=atol):
+        SU *= det ** (-1 / n)
+        global_phase = np.angle(det)
+
+    # Decompose the matrix
+    parameters_no_modes = _sun_parameters(SU, rtol, atol)
+
+    # Add the info about which modes each transformation is on
+    param_idx = 0
+    for md2 in range(2, n + 1):
+        for md1 in range(n - 1, md2 - 2, -1):
+            parameters.append([(md1 - 1, md1), parameters_no_modes[param_idx]])
+            param_idx += 1
+
+    return parameters, global_phase
+
+
+def _sun_parameters(U, rtol=1e-12, atol=1e-12):
```

Comment:
Just a note that this will be hidden in the docs, since it starts with an underscore. If the details in the docstring below is something that should be visible to the user, it either needs to be moved to the public function above or this function would need to be public.

Reply:
Same thing with a few functions below as well.

Reply:
I think in general the user does not need to know the gritty details of the decomposition but having the docstring here helps to understand the inner workings if someone whishes to do so.